### PR TITLE
core/vm: improve JUMPDEST analysis

### LINKED ISSUE #1147: JUMPDEST vulnerability
This test: https://github.com/ethereum/tests/blob/develop/StateTests/stSpecialTest.json#L2 has a transaction which costs only about 80k, but needs about 2GB of RAM and 18s in go.
The reason is the usage of a map to store JUMPDEST positions. In this test the code does 1024 recursive calls to itself. The code looks like: <1024 recursiv calls> STOP JUMPDEST JUMPDEST .... JUMPDEST (having 15k JUMPDESTS opcodes).
Cpp used a set for storing the JUMPDESTs, and needed 1.2 GB of RAM, we fixed it by using a vector of uint64_t instead (https://github.com/ethereum/cpp-ethereum/pull/2032), although this slightly reduces the performance.
This was found by a bug hunter through the bounty program. Gustav knows the details.

### PR DESCRIPTION
**Pull request by fjl**
* JUMPDEST analysis is faster because less type conversions are performed.
* The map of JUMPDEST locations is now created lazily at the first JUMP.
* The result of the analysis is kept around for recursive invocations
  through CALL/CALLCODE.

Fixes #1147

### CODE REVIEW

File: core/vm/analysis.go [hunk 1]
```diff
@@ -3,34 +3,43 @@ package vm
...
+	_, ok := m[dest.Uint64()]
+	return ok
 }
 
-func analyseJumpDests(code []byte) (dests *destinations) {
-	dests = &destinations{set.New()}
-
+// jumpdests creates a map that contains an entry for each
+// PC location that is a JUMPDEST instruction.
+func jumpdests(code []byte) map[uint64]struct{} {
```
Comment (Gustav-Simonsson) [inline core/vm/analysis.go]: Out of curiosity, why use a struct{} and not bool?

**Comment by Gustav-Simonsson** [comment]
Good to merge. @CJentzsch all tests pass with this, OK with you?

File: core/vm/analysis.go (hunk 1 above)
Comment (fjl) [inline core/vm/analysis.go]: Because struct{} has size zero.

**Comment by CJentzsch** [comment]
Looks good.

**Comment by fjl** [comment]
Jeff will look at it and merge when he is back from vacation.

File: core/vm/analysis.go [hunk 2]
```diff
@@ -3,34 +3,43 @@ package vm
...
+// destinations stores one map per contract (keyed by hash of code).
+// The maps contain an entry for each location of a JUMPDEST
+// instruction.
+type destinations map[common.Hash]map[uint64]struct{}
 
-func (d *destinations) Add(dest *big.Int) {
-	d.set.Add(string(dest.Bytes()))
+// has checks whether code has a JUMPDEST at dest.
+func (d destinations) has(codehash common.Hash, code []byte, dest *big.Int) bool {
+	if dest.Cmp(bigMaxUint64) > 0 {
```
Comment (obscuren) [inline core/vm/analysis.go]: requires documentation

**Comment by obscuren** [comment]
Looks good apart from the uint max dox

**Pull request merged**